### ISSUE TITLE
Suggestion for change in MetricWrapper

### ISSUE THREAD

**Issue by Seilmast** [SFI-Visual-Intelligence/Collaborative-Coding-Exam#64]
After doing some testing I've come to realise that the `MetricWrapper` class does not work well with metrics such as F1, Recall, and Percision. 
I'm suggesting handling all metric calculations within the metric classes themselves. I'll make a pull request draft and link it in a comment below. 

I'm suggesting then that the `__call__` or `forward` method of each class stores values needed to calculate the metric, then a method called `__getmetrics__` returns the metric currently stored and a `__reset__` method makes the instance ready for a new round of evaluation. 

The proposed MetricWrapper will look something like this: 

`class MetricWrapper(nn.Module):

    def __init__(self, num_classes, *metrics):
        super().__init__()
        self.metrics = {}
        self.num_classes = num_classes
        for metric in metrics:
            self.metrics[metric] = self._get_metric(metric)

    def _get_metric(self, key):
        match key.lower():
            case "entropy":
                return EntropyPrediction()
            case "f1":
                return F1Score(num_classes=self.num_classes)
            case "recall":
                return Recall(num_classes=self.num_classes)
            case "precision":
                return Precision(num_classes=self.num_classes)
            case "accuracy":
                return Accuracy(num_classes=self.num_classes)
            case _:
                raise ValueError(f"Metric {key} not supported")

    def __call__(self, y_true, y_pred):
        for key in self.metrics:
            self.metrics[key](y_true, y_pred)

    def __getmetrics__(self, str_prefix: str = None):
        return_metrics = {}
        for key in self.metrics:
            if str_prefix is not None:
                return_metrics[str_prefix + key] = self.metrics[key].__returnmetric__()
            else:
                return_metrics[key] = self.metrics[key].__returnmetric__()
        return return_metrics

    def reset(self):
        for key in self.metrics:
            self.metrics[key].reset()
`

**Comment by Seilmast** [comment]
See #65 for the suggested change. Ignore the formatting error...

**Comment by hzavadil98** [comment]
That is the issue of averaging the metrics batch-wise or dataset-wise you are referring to right? @Seilmast

**Comment by Seilmast** [comment]
Not sure I understand exactly what you're refering to, but I propose this change when I saw the F1 score not returning batchwise metrics. It might be difficult to do so for F1 score, and maybe precision / recall, so I'm saying we should maybe contain all metric calculations within the classes themselves.

**Comment by hzavadil98** [comment]
What I meant is that you can either calculate F1 metric separately for each batch and at the end of epoch just average over all the batches or accumulate the TP, FP, TN.. from all the batches and then calculate F1 only at the end of the epoch (which seems like the better option to me).

I think I am way too lost in all the changes. We should meet on thursday and merge all the changes and figure who needs to correct what in the 'personal' files.

**Comment by Seilmast** [comment]
Yes. There is a lot going on and we need to clear it up.
I made the proposed changes as I'm not a 100% sure you can do batchwise F1 score. I might be completely wrong on that front though.
The change I proposed would be to only calculate the metrics when calling __returnmetric()__. 
It doesnt matter for my metric really, but it seemed like a better solution in my head
